connect: Allow CA Providers to store small amount of state

banks opened the pull request:
**Merge Chain:** Note this is diffed against #6556 since it's continuing that work.

In preparation to adding new CA providers, there are cases where the configuration provided by the user is not sufficient for a CA provider to function as expected.

For example AWS PCA Provider needs to be able to create new root or subordinate CAs on first run (or forced rotation). When it does so the Unique ID (ARN) for that CA can be used in-memory to then reference that CA when making further API calls, e.g. to Sign certificates.

But on restart or leader failover, the new leader has only the user config (which had no ARN since they relied on the Provider to create one) so the only option would be to list all CAs visible in the PCA account and try to work out which is the right one based on cluster ID or common name. But that is ambiguous and assumes a lot about how we format Common Names etc. (particularly that they are unique enough for this case). Since the CAProvider doesn't have a unique identifier in Consul (the same provider instance might be used to manage multiple different Roots for example during a rotation) there is no way for Consul to pass an unambiguous ID, even if it could, it would prevent the provider from using it's own logic to decide which config changes require rolling a new key/cert etc. and which don't.

Much cleaner is to allow CA Providers to optionally return some arbitrary state in form of key/value strings. I made this a new method that Consul will call after ever CA config change once it needs to persist and the output if any is persisted to the state store along with the CA config.

The current state persisted is then provided to subsequent `Configure` calls, provided the provider being configured is the same provider type.

This will allow the AWS provider to create new CAs and return the ARN from it's `State` method but be able to unambiguously load the resource and continue to sign with the same CA after a restart/leader election.

rboyer commented on agent/structs/connect_ca.go:
```diff
@@ -366,6 +426,16 @@ type VaultCAProviderConfig struct {
 	TLSSkipVerify bool
 }
 
+type AWSCAProviderConfig struct {
```
I suspect you meant to keep this in your other PR.

rboyer commented on agent/connect_ca_endpoint.go:
```diff
@@ -72,26 +71,3 @@ func (s *HTTPServer) ConnectCAConfigurationSet(resp http.ResponseWriter, req *ht
-// so that they get formatted correctly during json.Marshal. Without this,
-// string values that get converted to []uint8 end up getting output back
-// to the user in base64-encoded form.
-func fixupConfig(conf *structs.CAConfiguration) {
```
Nice to see this move over into a more "standard" place (the Marshal/Unmarshal functions).

rboyer commented on agent/consul/connect_ca_endpoint.go:
```diff
@@ -168,6 +168,13 @@ func (s *ConnectCA) ConfigurationSet(
+	// so it can decide if it needs to change resources or not based on the config
+	// change.
+	if args.Config.Provider == config.Provider {
+		args.Config.State = config.State
```
Should this do a deep copy?

banks replied:
Good question. I guess I should document that in the interface. I'm leaning towards no given that it's a `map[string]string` not a blind pointer to something the Provider is likely to hold in the struct but we can document that assumption. If the provider is going to store a map and return it then _it_ should do the copy. In most cases I suspect `State` method will be building this map based on other actual struct members anyway.

mkeeler commented on agent/connect/ca/plugin/provider.proto:
```diff
@@ -49,6 +51,10 @@ message CrossSignCARequest {
 }
 
+message StateResponse {
+    bytes state = 1; // JSON-encoded map[string]string
```
Do we enforce that its JSON-encoded. Given the PR title this seems like its intended for opaque data that is encoded somehow.

banks commented:
Changed base to master after merging the parent branch and fixed conflicts. Will wait to see if CI passes (primary affected packages all passed locally).
